### PR TITLE
LPS-187576 |Functional Automation - AccessibilityMenu#LinkIsKeyboardFocusable and AccessibilityMenu#IsKeyboardAccessible

### PR DESCRIPTION
**Pull request by eltonccdantas**
Create the functional automation for the test scenarios that is not possible to cover in a backend or frontend side. The priority of the test scenarios must be critical, high, or mid.

Create a subtask for each testcase

Assign subtask to yourself

Send to liferay-peds and request QA reviewer

Close subtask when merged

Test Scenarios
Prerequisites:

Deploy accessibility-menu-web

Enable Accessibility Menu in Instance Settings > Accessibility

- [x] AccessibilityMenu#LinkIsKeyboardFocusable
- [x] AccessibilityMenu#IsKeyboardAccessible

[Jira](https://liferay.atlassian.net/browse/LPS-186427)

### CODE REVIEW

File: modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase [hunk 1]
```diff
@@ -26,10 +47,15 @@ definition {
 		if (${testPortalInstance} == "true") {
 			PortalInstances.tearDownCP();
 
+			TestUtils.hardRefresh();
+
 			// Add clear browser cache
```
Comment (eltonccdantas) [inline modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase:52]: I'll remove it

File: modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase [hunk 2]
```diff
@@ -40,215 +66,246 @@ definition {
...
-			// And When
-			// guest user toggles show underline to ON
-			// And guest user toggles show underline to OFF
-			// And sign in as user A
-			// Then show underline is set to ON state
 
-		}
+		//task ("Given When Then") {
+
+		// // Given a new created user A
```
Comment (eltonccdantas) [inline modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase:72]: I needed to comment the task in order to run the poshi tests locally

File: modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase [hunk 3]
```diff
@@ -40,215 +66,246 @@ definition {
...
+				value1 = "\ENTER");
+		}
 
+		task ("Then Accessibility Menu modal is present") {
+			AssertElementPresent(
+				key_modal_title = "Accessibility Help Menu",
+				locator1 = "AccessibilityMenu#MODAL_TITLE");
+		}
+
+		task ("And can navigate to a switch toggle with keyboard") {
```
Comment (eltonccdantas) [inline modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase:241]: I think I'll create a macro for it, depending on the other tests

File: modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase [hunk 4]
```diff
@@ -40,215 +66,246 @@ definition {
...
-			//TestUtils.assertElementFocused();
-
+			TestUtils.assertElementFocused(element = "//*[@aria-label='Quick Links']//li//*[contains(text(),'Open Accessibility Menu')]");
 		}
 	}
 
 	@description = "LPS-178192. Verifies accessibility menu modal can be accessed by keyboard."
 	@ignore = "Test Stub"
-	@priority = 3
+	@priority = 5
```
Comment (eltonccdantas) [inline modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase:222]: Adjusted the priority according to the ticket

**Comment by john-co** [comment]
Just started reviewing :)

File: modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase [hunk 5]
```diff
@@ -26,10 +47,15 @@ definition {
 		if (${testPortalInstance} == "true") {
 			PortalInstances.tearDownCP();
 
+			TestUtils.hardRefresh();
+
```
Comment (john-co) [inline modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase:51]: Refreshing the page doesn't clear the browser cache.

File: modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase [hunk 6]
```diff
@@ -26,10 +47,15 @@ definition {
...
 			// Add clear browser cache
 
 		}
 		else {
+			TestUtils.hardRefresh();
+
+			PortalInstances.tearDownCP();
```
[the remark this answers is not in the document]
Comment (john-co) [reply modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase:58]: same comment about refresh. 

PortalInstances.tearDownCP(); is not needed here b/c portal instances would be false at this point.

File: modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase [hunk 7]
```diff
@@ -40,215 +66,246 @@ definition {
...
 		task ("When press tab key from 'Skip to Content' link") {
-
-			// Example:
-
-			//KeyPress(
-			//	key_text = "New Experience",
-			//	locator1 = "Button#ANY",
-			//	value1 = "\TAB");
-
+			AccessibilityMenu.goAccessibilityMenu();
```
Comment (john-co) [inline modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase:212]: We want to specifically start from `Skip to Content` link, not from the top of the page/body.

File: modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase [hunk 8]
```diff
@@ -40,215 +66,246 @@ definition {
...
 		task ("Then Accessibility Menu link is focusable") {
-
-			//TestUtils.assertElementFocused();
-
+			TestUtils.assertElementFocused(element = "//*[@aria-label='Quick Links']//li//*[contains(text(),'Open Accessibility Menu')]");
 		}
 	}
 
 	@description = "LPS-178192. Verifies accessibility menu modal can be accessed by keyboard."
 	@ignore = "Test Stub"
```
Comment (john-co) [inline modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase:221]: Remove the ignore line

File: modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase [hunk 9]
```diff
@@ -40,215 +66,246 @@ definition {
...
+			KeyPress(
+				locator1 = "//body",
+				value1 = "\TAB");
+
+			KeyPress(
+				key_modalTitle = "Accessibility Help Menu",
+				locator1 = "Button#CLOSE_MODAL",
+				value1 = "\TAB");
+
+			TestUtils.assertElementFocused(element = "//label[contains(.,'Underlined Links')]//input[@type='checkbox' and not(@checked)]");
```
Comment (john-co) [inline modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase:251]: ok for now, need to update this after new poshi function is implemented: https://liferay.slack.com/archives/CD7939WBE/p1686950932162979

File: modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase [hunk 10]
```diff
@@ -5,18 +5,39 @@ definition {
...
+					checkboxName = "Enable Accessibility Menu",
+					locator1 = "Checkbox#ANY_CHECKBOX");
+
+				if (IsElementPresent(locator1 = "Button#SAVE")) {
+					Click(locator1 = "Button#SAVE");
+				}
+				else {
+					Click(locator1 = "Button#UPDATE");
+				}
+			}
```
Comment (john-co) [inline modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase:36]: Create a macro for this

**Review by john-co — CHANGES_REQUESTED** [review]

File: modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase (hunk 5 above)
Comment (john-co) [reply modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase:51]: You may want to try closing the browser window and opening incognito window.

File: modules/apps/accessibility/accessibility-test/src/testFunctional/macros/AccessibilityMenu.macro [hunk 11]
```diff
@@ -0,0 +1,17 @@
+definition {
+
+	macro goAccessibilityMenu {
+		KeyPress(
+			locator1 = "//body",
+			value1 = "\TAB");
+
+		KeyPress(
+			locator1 = "//body",
+			value1 = "\TAB");
```
[the remark this answers is not in the document]
Comment (eltonccdantas) [reply modules/apps/accessibility/accessibility-test/src/testFunctional/macros/AccessibilityMenu.macro:10]: Yeah...  `//body` worked fine for it

File: modules/apps/accessibility/accessibility-test/src/testFunctional/paths/AccessibilityMenu.path [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+
+<tr>
+	<td>QUICK_LINK</td>
+	<td>//*[@aria-label="Quick Links"]//li//*[contains(text(),"${key_link}")]</td>
+	<td></td>
+</tr>
+<tr>
+	<td>MODAL_TITLE</td>
+	<td>//*[@class="modal-title"][contains(text(),"${key_modal_title}")]</td>
+	<td></td>
```
[the remark this answers is not in the document]
Comment (eltonccdantas) [reply modules/apps/accessibility/accessibility-test/src/testFunctional/paths/AccessibilityMenu.path:22]: Nope... but I could try

File: modules/apps/accessibility/accessibility-test/src/testFunctional/paths/AccessibilityMenu.path [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+</tr>
+<tr>
+	<td>MENU_TOGGLE</td>
+	<td>//*[@class="toggle-switch simple-toggle-switch"]//span[contains(text(),"${key_toggle_label}")]</td>
+	<td></td>
```
[the remark this answers is not in the document]
Comment (eltonccdantas) [reply modules/apps/accessibility/accessibility-test/src/testFunctional/paths/AccessibilityMenu.path:27]: I think I tried this one... but I can review it

File: modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase [hunk 14]
```diff
@@ -5,18 +5,39 @@ definition {
 	property portal.release = "false";
 	property portal.upstream = "true";
 	property testray.component.names = "Accessibility Menu";
-	property testray.main.component.name = "Accessibility Menu";
+	property testray.main.component.name = "Accessibility Framework";
```
[the remark this answers is not in the document]
Comment (eltonccdantas) [reply modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase:8]: It didn't work... the Poshi test didn't even start

[the remark this answers is not in the document]
Comment (john-co) [reply modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase:8]: I'll look into it further, I think we have to add this component name somewhere.

File: modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase (hunk 5 above)
Comment (eltonccdantas) [reply modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase:51]: Right... I thought that hard refresh would also empty cash, thanks for clarifying...

File: modules/apps/accessibility/accessibility-test/src/testFunctional/paths/AccessibilityMenu.path (hunk 12 above)
[the remark this answers is not in the document]
Comment (eltonccdantas) [reply modules/apps/accessibility/accessibility-test/src/testFunctional/paths/AccessibilityMenu.path:22]: It did not work with `IFrame#IFRAME_TITLE`

File: modules/apps/accessibility/accessibility-test/src/testFunctional/paths/AccessibilityMenu.path (hunk 13 above)
[the remark this answers is not in the document]
Comment (eltonccdantas) [reply modules/apps/accessibility/accessibility-test/src/testFunctional/paths/AccessibilityMenu.path:27]: I realized I didn't even use it... I'll wipe it out.

File: modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase (hunk 5 above)
Comment (eltonccdantas) [reply modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase:51]: [Please see it](https://liferay.slack.com/archives/CD7939WBE/p1648168045484689?thread_ts=1648061188.934189&cid=CD7939WBE)

**Comment by eltonccdantas** [comment]
ci:test:sf

**Comment by eltonccdantas** [comment]
ci:test:relevant

**Comment by eltonccdantas** [comment]
ci:reevaluate:1387401_8118

**Comment by eltonccdantas** [comment]
ci:test:relevant

**Comment by eltonccdantas** [comment]
ci:forward:force

**Comment by eltonccdantas** [comment]
The [failures look unrelated](https://testray.liferay.com/home/-/testray/runs/compare?compare=true&view=details&statusA=2&testrayRunIdB=2366478565&testrayRunIdA=2366214663&statusB=3) to it

**Comment by eltonccdantas** [comment]
ci:forward:force

**Comment by eltonccdantas** [comment]
Manually forwarded to https://github.com/brianchandotcom/liferay-portal/pull/137206

**Comment by john-co** [comment]
I want to review final version of this before forwarding.

File: modules/apps/accessibility/accessibility-test/src/testFunctional/macros/AccessibilityMenu.macro [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+				Click(locator1 = "Button#SAVE");
+			}
+			else {
+				Click(locator1 = "Button#UPDATE");
+			}
+		}
+
+		task ("And navigate to home page") {
+			Navigator.openURL();
+		}
```
Comment (john-co) [inline modules/apps/accessibility/accessibility-test/src/testFunctional/macros/AccessibilityMenu.macro:28]: We usually keep the task annotations in testcase files.

File: modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase (hunk 14 above)
[the remark this answers is not in the document]
Comment (john-co) [reply modules/apps/accessibility/accessibility-test/src/testFunctional/tests/AccessibilityMenu.testcase:8]: This wasn't resolved yet so I'm opening this thread back up for now.

**Review by john-co — COMMENTED** [review]
Looks good! I'm going to include some minor updates

**Comment by john-co** [comment]
ci:test:accessibility-menu:nocompile

**Comment by john-co** [comment]
ci:test:sf

**Comment by john-co** [comment]
ci:test:accessibility-menu:nocompile

**Comment by john-co** [comment]
ci:test:sf

**Comment by john-co** [comment]
ci:test:stable

**Comment by john-co** [comment]
ci:test:relevant

**Comment by john-co** [comment]
ci:test:accessibility-menu:nocompile

**Comment by john-co** [comment]
Hi @eltonccdantas, the test [LocalFile.AccessibilityMenu#LinkIsKeyboardFocusable](https://testray.liferay.com/home/-/testray/case_results/2377003671?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D2376999797%26testrayRunId%3D2377003597%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc) fails on CI. Can you look into this further to stabilize?

**Comment by eltonccdantas** [comment]
> Hi @eltonccdantas, the test [LocalFile.AccessibilityMenu#LinkIsKeyboardFocusable](https://testray.liferay.com/home/-/testray/case_results/2377003671?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D2376999797%26testrayRunId%3D2377003597%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc) fails on CI. Can you look into this further to stabilize?

Sure... I'll work on it

**Comment by eltonccdantas** [comment]
ci:test:relevant

**Comment by eltonccdantas** [comment]
ci:test:sf

**Comment by eltonccdantas** [comment]
ci:test:sf

**Comment by eltonccdantas** [comment]
ci:test:sf

**Comment by eltonccdantas** [comment]
ci:test:relevant

**Comment by eltonccdantas** [comment]
ci:test:accessibility-menu

**Comment by john-co** [comment]
ci:test:accessibility-menu

**Comment by john-co** [comment]
ci:test:sf

**Comment by john-co** [comment]
ci:test:relevant

**Comment by john-co** [comment]
@eltonccdantas , looks like the test is still failing.

**Comment by eltonccdantas** [comment]
ci:test:sf

**Comment by eltonccdantas** [comment]
ci:test:accessibility-menu

**Comment by eltonccdantas** [comment]
I realized that sometimes poshi executes the steps so fast that the accessibility menu was not even enabled, so I added verification for the alert toast in order to to take "longer" before the next step. Let's try

**Comment by eltonccdantas** [comment]
image

**Comment by eltonccdantas** [comment]
ci:test:relevant

**Comment by eltonccdantas** [comment]
Hey @john-co, looks good to me... What do you think? I'll keep waiting on your final review before forwarding it

**Comment by eltonccdantas** [comment]
ci:test:relevant

**Comment by eltonccdantas** [comment]
ci:test:relevant

**Comment by john-co** [comment]
ci:test:stable

**Comment by eltonccdantas** [comment]
Hey @john-co, may we forward this one?

**Comment by john-co** [comment]
looks good to me! :sunglasses:

**Comment by john-co** [comment]
ci:forward:force

**Comment by eltonccdantas** [comment]
ci:forward:force

**Comment by eltonccdantas** [comment]
Manually forwarded to https://github.com/brianchandotcom/liferay-portal/pull/138026

**Pull request closed without merging**